# 1. Importing Libraries and DFs

In [2]:
import pandas as pd
import numpy as np
import os

In [3]:
path = r'C:\Users\Marvin\Instacart Basket Analysis'

In [4]:
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'),index_col = False)

In [5]:
df_ords = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'),index_col = False)

# 2. Mixed-type Data

In [7]:
# Create a dataframe
df_test = pd.DataFrame()

In [8]:
# Create a mixed type column
df_test['mix'] = ['a','b',1,True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# Check for mixed values
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [11]:
# Fixing column
df_test['mix'] = df_test['mix'].astype('str')

# 3. Missing Values

In [12]:
# Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
# Creating DF with Null values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
df_prods.shape

(49693, 5)

In [16]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
df_prods_clean.shape

(49677, 5)

# 4. Duplicates

In [18]:
# Finding Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [20]:
df_prods_clean.shape

(49677, 5)

In [21]:
# Drop duplicates in new DF
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dups.shape

(49672, 5)

In [23]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data','products_checked.csv'))

# Task 4.5

## 2. Data Consistency Check

In [32]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In the df_prods data there is an outlier in the prices column, where the max value is 99999 which would be way to high, while the 50% is only 7.1.
It's very likely that this was a typo.

In [24]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

        time_of_day  days_since_prior_order  
count  3.421083e+06            3.214874e+06  
mean   1.345202e+01            1.111484e+01  
std    4.226088e+00            9.206737e+00  
min    0.000000e+00            0.000000e+00  
25%    1.000000e+01            4.000000e+00  
50%    1.300000e+01            7.000000e+00  
75%    1.600000e+01            1.500000e+01  
max    2.300000e+01            3.000000e+01

Investigating the frequency table in the max and min section, it shows that the days_since_prior_order has weird recordings.
The min is 0 and the max is 3, while the 50th percentile is 7, which is higher.
Also the time_of_day max only goes until 23, which should be until 24.

## 3. Check for mixed-type Data

In [25]:
# Check for mixed values
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
else:
    print('No mixed values were found')

No mixed values were found


## 4. No mixed type data found

## 5. Checking for missing values

In [26]:
# Finding missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_dow                      0
time_of_day                    0
days_since_prior_order    206209
dtype: int64

In the days_since_prior_order column are 206209 missing values

## 6. Missing values and appropriate method

In [27]:
df_ords.shape

(3421083, 7)

Because 206209 missing values is a big chunk of data that is missing from 3.4 Mio. rows, I would not do anything with the missing values. It wouldn't be safe to delete the data as this could lead to wrong analysis.

## 7. Checking for duplicates

In [28]:
# Finding Duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [29]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_dow, time_of_day, days_since_prior_order]
Index: []

No duplicates were found

## 8. Duplicate Method

No duplicate method need to be used, because there were no duplicates found in the dataframe.

## 9. Exporting Data

In [30]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data','orders_checked.csv'))

In [31]:
df_prods.to_csv(os.path.join(path,'02 Data','Prepared Data','products_checked'))